### DISCUSSION TITLE
Converting to juypter doesn't display in output of notebook

### DISCUSSION THREAD

**Discussion by sherodtaylor** [marimo-team/marimo#1670] [General]
### Describe the bug

After exporting to jupyter notebook running markdown or any other output doesn't display in notebook. 

<img width="1608" alt="Screenshot 2024-06-24 at 10 49 46 AM" src="https://github.com/marimo-team/marimo/assets/1483092/15d0799c-0331-48fe-a144-6366a0c23d49">

Also if you try to do `mo.output.append` it throws a no context error. I'd like to be able to convert to a jupyter notebook book for a regularly scheduled report my company uses.

### Environment

```
{
  "marimo": "0.6.19",
  "OS": "Darwin",
  "OS Version": "23.5.0",
  "Processor": "arm",
  "Python Version": "3.11.6",
  "Binaries": {
    "Browser": "126.0.6478.63",
    "Node": "v20.10.0"
  },
  "Requirements": {
    "click": "8.1.7",
    "importlib-resources": "missing",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.8.1",
    "pygments": "2.17.2",
    "tomlkit": "0.12.5",
    "uvicorn": "0.30.1",
    "starlette": "0.37.2",
    "websockets": "12.0",
    "typing-extensions": "4.9.0",
    "black": "24.4.2"
  }
}
```

### Code to reproduce

```python
import marimo

__generated_with = "0.6.19"
app = marimo.App()


@app.cell
def __():
    import marimo as mo

    mo.md("""
    output should display in ipython
    """)
    return mo,


@app.cell
def __(mo):
    mo.output.append(mo.md("hello"))
    return


if __name__ == "__main__":
    app.run()

```

**Comment by mscolnick** [comment]
I am converting this to a discussion, as it's not quite an issue for us.

marimo is not built on top of Jupyter so I wouldn't expect it to run in Jupyter at all if you are using the marimo library (`import marimo as mo`). The marimo library is very integrated with the marimo kernel - which allows us to do a lot of cool things that Jupyter otherwise cannot: reactive ui elements, interpolated markdown, lazy-loading, layouting, streaming outputs, etc.

In order to improve upon the notebook and not be help back by legacy features, technologies, or specs, it is a non-goal to allow the marimo library `import marimo as mo` to full work inside of Jupyter.

**Comment by sherodtaylor** [comment]
i agree with the premise of not fully supporting output but why support exporting as a feature if it can't run. Maybe it shouldn't be a bug but semi-feature to allow more people to migrate off of it. 

i agree you shouldn't couple it to jupyter but it would be trivial enough to support writing the output or ipython display
it actually does run to some extent it would be nice if you supported a basic output

you could literally just check if it is running in a juypter notebook and call ipython display under the hood. 

https://stackoverflow.com/questions/15411967/how-can-i-check-if-code-is-executed-in-the-ipython-notebook

**Reply by akshayka** [reply]
@sherodtaylor

> you could literally just check if it is running in a juypter notebook and call ipython display under the hood.

We'd welcome a pull request for that. We're not intentionally not supporting things. Just limited on time.

**Reply by sherodtaylor** [reply]
thats fair i just thought i'd raise this as it advertises convert from and to jupyter notebooks.

**Comment by mscolnick** [comment]
Printing to ipython will likely create more bugs than solve. For example, printing a slider will just return a non-working slider that doesn't update any downstream code. Our tables and charts use some semi-fancy lazy-loading for performance, so that would need to be ported over as well.

> why support exporting as a feature

I agree with this sentiment, and it makes it feel like the export is half-done. I think adding some level out python outputs to jupyter will make this feel similarly half-done (i.e. no interactions work, design is off, etc)

We did this originally because it is a respected format that can serve as a bridge to other systems or CLIs (GitHub previews, pandoc, myst, jupytext, etc). For example, `marimo export ipynb > some_other_cli`

**Reply by sherodtaylor** [reply]
> I agree with this sentiment, and it makes it feel like the export is half-done. I think adding some level out python outputs to jupyter will make this feel similarly half-done (i.e. no interactions work, design is off, etc)

Yeah thats fair but I think you could probably put in a section around interactivity not being a supported like you note with `matplotlib` interactions.  If you are supporting exports, outputting in ipython is a fair ask. It wouldn't feel half done as you clearly state the support of it.

> We did this originally because it is a respected format that can serve as a bridge to other systems or CLIs (GitHub previews, pandoc, myst, jupytext, etc). For example, marimo export ipynb > some_other_cli

A bridge to runnable ipynb makes it appealing to companies that have tooling around jupyter notebooks. For example we have a tooling around scheduled reports via notebooks that we'd love to use but would also like to use this library. The reproducibility is the most appealing for us.

**Reply by mscolnick** [reply]
markdown is something we can look into supporting without much lift. 

the others need web-components installed so those won't actually render in Jupyter anyways, without some custom extension

**Reply by sherodtaylor** [reply]
thats good to know! Could you support pure html like if we were using `matplotlib.pyplot` or `seaborn`?

**Reply by mscolnick** [reply]
Does that not work today? Or is it that we don't include the generated html in the notebook json?

**Reply by sherodtaylor** [reply]
it doesn't work. It just prints the object and not the generated html

**Reply by mscolnick** [reply]
including when using `matplotlib.pyplot` and `seaborn`?

**Reply by sherodtaylor** [reply]
@mscolnick actually just checked it is outputting it correctly it's only when using mo.as_html or mo.md

**Reply by sherodtaylor** [reply]
^ i haven't found a way to get the raw html from the mo.Html object.

**Reply by akshayka** [reply]
@sherodtaylor, use the `text` attribute: https://docs.marimo.io/api/html.html#marimo.Html.text
